src/tron/tronEngine.js

```diff
@@ -0,0 +1,548 @@
+
+  async startEngine () {
+    this.engineOn = true
+    this.addToLoop('checkBlockchainInnerLoop', BLOCKCHAIN_POLL_MILLISECONDS)
+    this.addToLoop('checkAccountInnerLoop', ACCOUNT_POLL_MILLISECONDS)
+    // this.addToLoop('checkUpdateNetworkFees', NETWORKFEES_POLL_MILLISECONDS)
+    this.addToLoop('checkTransactionsInnerLoop', TRANSACTION_POLL_MILLISECONDS)
+    super.startEngine()
+  }
+
+  async resyncBlockchain () {
+    await this.killEngine()
+    await this.clearBlockchainCache()
+    await this.startEngine()
+  }
+
+  async makeSpend (edgeSpendInfoIn: EdgeSpendInfo) {
+    try {
+      const { edgeSpendInfo, currencyCode } = super.makeSpend(edgeSpendInfoIn)
+      const networkFee = '0'
+      const nativeAmount = edgeSpendInfo.spendTargets[0].nativeAmount
+      const balanceTrx = this.walletLocalData.totalBalances[
+        this.currencyInfo.currencyCode
+      ]
+
+      if (bns.gt(nativeAmount, balanceTrx)) {
+        throw new InsufficientFundsError()
+      }
+
+      // Tron can only have one output
+      if (edgeSpendInfo.spendTargets.length !== 1) {
+        throw new Error('Error: only one output allowed')
+      }
+
+      const spendTarget = edgeSpendInfo.spendTargets[0]
+      const publicAddress = spendTarget.publicAddress
+      const data =
+        spendTarget.otherParams != null ? spendTarget.otherParams.data : void 0
+
+      let otherParams: Object = {}
+      // if TRON
+      if (currencyCode === PRIMARY_CURRENCY) {
+        // builds tx, but does NOT send it
+        const trxParams: TronTxOtherParams = await tronWeb.transactionBuilder.sendTrx(
+          publicAddress,
+          nativeAmount,
+          this.walletLocalData.publicKey
+        )
+        otherParams = trxParams
+        // if Token
+      } else {
+        let contractAddress = ''
+        if (data) {
+          contractAddress = publicAddress
+        } else {
+          const tokenInfo = this.getTokenInfo(currencyCode)
+          if (!tokenInfo || typeof tokenInfo.contractAddress !== 'string') {
+            throw new Error(
+              'Error: Token not supported or invalid contract address'
+            )
+          }
+
+          contractAddress = tokenInfo.contractAddress
+        }
+
+        // for Tokens:
+        // builds tx, but does NOT send it
+        const trxParams: TronTxOtherParams = await tronWeb.transactionBuilder.sendToken(
+          publicAddress,
+          nativeAmount,
+          contractAddress,
+          this.walletLocalData.publicKey
+        )
+        otherParams = trxParams
+      }
+      const edgeNativeAmount = bns.mul(nativeAmount, '-1')
+
+      // **********************************
+      // Create the unsigned EdgeTransaction
+
+      const edgeTransaction: EdgeTransaction = {
+        txid: '', // txid
+        date: 0, // date
+        currencyCode, // currencyCode
+        blockHeight: 0, // blockHeight
+        nativeAmount: edgeNativeAmount, // nativeAmount
```

Comment:
Native amount should include the network fee. Network fee is a constant above that is set to zero. Is this really true of Tron? I was not aware that Tron has zero fees.

Reply:
Should be fine. If you need any code changes, let me know. From the docs:

The TRON network grants every account a free pool of Bandwidth Points for free transactions every 24 hours. To engage in transactions more frequently requires freezing TRX for additional bandwidth points, or paying the fee in TRX.

There are 5000 bandwidth points for free per account per day. When an account hasn’t frozen any balance, or when its bandwidth points have run out, complimentary bandwidth points can be used. Each transaction in the TRON network is about 200 bytes, so each account enjoys about 25 transactions for free each day.

If ran out of free transactions costs are:
For a normal transfer, it costs about 0.002 TRX (USD 0.000028)

Activating an account costs 0.1TRX (USD 0.0014)

1 TRX = 0.014$

https://developers.tron.network/docs/bandwith